source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc

```diff
@@ -8,10 +10,92 @@ namespace Extensions {
 namespace Tracers {
 namespace OpenTelemetry {
 
+static constexpr std::chrono::seconds INITIAL_TIMER_DURATION{10};
+static constexpr std::chrono::minutes TIMER_INTERVAL{5};
+
+namespace {
+
+bool reEnableTimer(Http::Code response_code) {
+  switch (response_code) {
+  case Http::Code::OK:
+  case Http::Code::TooManyRequests:
+  case Http::Code::InternalServerError:
+  case Http::Code::BadGateway:
+  case Http::Code::ServiceUnavailable:
+  case Http::Code::GatewayTimeout:
+    return true;
+  default:
+    return false;
+  }
+}
+
+} // namespace
+
 SamplerConfigProviderImpl::SamplerConfigProviderImpl(
-    Server::Configuration::TracerFactoryContext& /*context*/,
+    Server::Configuration::TracerFactoryContext& context,
     const envoy::extensions::tracers::opentelemetry::samplers::v3::DynatraceSamplerConfig& config)
-    : sampler_config_(config.root_spans_per_minute()) {}
+    : cluster_manager_(context.serverFactoryContext().clusterManager()),
+      http_uri_(config.http_uri()),
+      authorization_header_value_(absl::StrCat("Api-Token ", config.token())),
+      sampler_config_(config.root_spans_per_minute()) {
+
+  timer_ = context.serverFactoryContext().mainThreadDispatcher().createTimer([this]() -> void {
+    const auto thread_local_cluster = cluster_manager_.getThreadLocalCluster(http_uri_.cluster());
+    if (thread_local_cluster == nullptr) {
+      ENVOY_LOG(error, "SamplerConfigProviderImpl failed: [cluster = {}] is not configured",
+                http_uri_.cluster());
+    } else {
+      Http::RequestMessagePtr message = Http::Utility::prepareHeaders(http_uri_);
+      message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Get);
+      message->headers().setReference(Http::CustomHeaders::get().Authorization,
+                                      authorization_header_value_);
+      active_request_ = thread_local_cluster->httpAsyncClient().send(
+          std::move(message), *this,
+          Http::AsyncClient::RequestOptions().setTimeout(std::chrono::milliseconds(
+              DurationUtil::durationToMilliseconds(http_uri_.timeout()))));
+    }
+  });
+
+  timer_->enableTimer(std::chrono::seconds(INITIAL_TIMER_DURATION));
+}
+
+SamplerConfigProviderImpl::~SamplerConfigProviderImpl() {
+  if (active_request_) {
+    active_request_->cancel();
+  }
+}
+
+void SamplerConfigProviderImpl::onSuccess(const Http::AsyncClient::Request& /*request*/,
+                                          Http::ResponseMessagePtr&& http_response) {
+  active_request_ = nullptr;
+  const auto response_code = Http::Utility::getResponseStatus(http_response->headers());
+  bool json_valid = false;
+  if (response_code == enumToInt(Http::Code::OK)) {
+    ENVOY_LOG(debug, "Received sampling configuration from Dynatrace: {}",
+              http_response->bodyAsString());
+    json_valid = sampler_config_.parse(http_response->bodyAsString());
```

Comment:
Will we access the `sampler_config_` at worker threads? If yes, then seems there maybe a contention.

Reply:
yes, `sampler_config` is used in worker threads. But it has only two members:
* one is `const` and never modified during runtime.
* the other one is set during `parse` and read in the worker threads. It is an `std::atomic<uint32_t>`, so it should be ok to access the variable in multiple threads.